### ISSUE TITLE
SFTP slow with many directories

### ISSUE THREAD

**Issue by jnordberg** [rclone/rclone#7746]
Using `rclone copy local/path sftp_remote:remote/path` when `local/path` contains 500k directories on a high latency connection takes forever since directories are created first (and and seemingly serially, ignoring the `--transfers` flag)

Using `--progress` makes it look like the transfer has stalled, but with `-vv` there I can see a stream of:

```
2024/04/09 09:02:23 DEBUG : [redacted]: Making directory
2024/04/09 09:02:24 DEBUG : [redacted]: Making directory
```

I've tried `--order-by size,desc` and `--inplace` but it seems determined to create all the directories before starting to transfer (which means I'm looking at 5 days just to start the transfer at this rate).

I've also tried to `ls -f local/path > dir_list.txt` and then copy that to the destination and run `while read -r line; do mkdir -p "$line"; done < dir_list.txt` to-pre populate the directories before running rclone but that doesn't help. After doing that it's still stuck and log is silent, only log for 30 minutes was:

```
2024/04/09 09:56:35 DEBUG : sftp://[redacted]: Closing 63 unused connections
```

I'm guessing it's doing some similarly slow and serial stat on all directories before starting.

---

```
$ rclone --version
rclone v1.66.0
- os/version: ubuntu 22.04 (64 bit)
- os/kernel: 5.15.130-0515130-generic (x86_64)
- os/type: linux
- os/arch: amd64
- go/version: go1.22.1
- go/linking: static
- go/tags: none
```

**Comment by ncw** [comment]
I think this is related to the fix going on here

https://github.com/rclone/rclone/issues/7689

So I suggest you try the beta in that issue. It isn't quite perfect yet but it should help.

**Comment by jnordberg** [comment]
I think this is different, I don't have any empty directories and I'm running with `--size-only --no-update-modtime --no-update-dir-modtime`.

The workaround with pre-creating directories seems to work though, I was just not patient enough, after roughly 1 hour of seemingly stalling it kicked off and started transferring files.

**Comment by ncw** [comment]
> I think this is different, I don't have any empty directories and I'm running with `--size-only --no-update-modtime --no-update-dir-modtime`.

The fix above stops rclone pre-creating the directories which also stops it creating empty directories it doesn't need.

> The workaround with pre-creating directories seems to work though, I was just not patient enough, after roughly 1 hour of seemingly stalling it kicked off and started transferring files.

:+1:

**Comment by edc-w** [comment]
closing issue as it looks complete @jnordberg please make a new issue if problem persists :-)
